lib/sparesults/src/json.rs

```diff
@@ -224,520 +220,882 @@ fn write_json_term<'a>(output: &mut Vec<JsonEvent<'a>>, term: TermRef<'a>) {
+            }
+            JsonInnerTermReaderState::Value => match event {
+                JsonEvent::String(value) => {
+                    self.value = Some(value.into_owned());
+                    self.state = JsonInnerTermReaderState::Middle;
+                    Ok(None)
+                }
+                #[cfg(feature = "rdf-star")]
+                JsonEvent::StartObject => {
+                    self.state = JsonInnerTermReaderState::InValue;
+                    Ok(None)
+                }
+                _ => {
+                    self.state = JsonInnerTermReaderState::Middle;
+
+                    Err(QueryResultsSyntaxError::msg("Term value must be a string"))
+                }
+            },
+            JsonInnerTermReaderState::Lang => {
+                let result = if let JsonEvent::String(value) = event {
+                    self.lang = Some(value.into_owned());
+                    Ok(None)
+                } else {
+                    Err(QueryResultsSyntaxError::msg("Term lang must be strings"))
+                };
+                self.state = JsonInnerTermReaderState::Middle;
+
+                result
+            }
+            JsonInnerTermReaderState::Datatype => {
+                let result = if let JsonEvent::String(value) = event {
+                    match NamedNode::new(value) {
+                        Ok(datatype) => {
+                            self.datatype = Some(datatype);
+                            Ok(None)
                         }
+                        Err(e) => Err(QueryResultsSyntaxError::msg(format!(
+                            "Invalid datatype: {e}"
+                        ))),
                     }
+                } else {
+                    Err(QueryResultsSyntaxError::msg("Term lang must be strings"))
+                };
+                self.state = JsonInnerTermReaderState::Middle;
+
+                result
+            }
+            #[cfg(feature = "rdf-star")]
+            JsonInnerTermReaderState::InValue => match event {
+                JsonEvent::ObjectKey(object_key) => {
+                    self.state = match object_key.as_ref() {
+                        "subject" => JsonInnerTermReaderState::Subject(Box::default()),
+                        "predicate" => JsonInnerTermReaderState::Predicate(Box::default()),
+                        "object" => JsonInnerTermReaderState::Object(Box::default()),
+                        _ => {
+                            return Err(QueryResultsSyntaxError::msg(format!(
+                                "Unsupported value key: {object_key}"
+                            )));
+                        }
+                    };
+                    Ok(None)
+                }
+                JsonEvent::EndObject => {
+                    self.state = JsonInnerTermReaderState::Middle;
+                    Ok(None)
                 }
-                _ => ignore_value(reader)?,
+                _ => unreachable!(),
```

Comment:
here and elsewhere i noticed that you don't have an exhaustive list of enums in a match -- there is even a [lint](https://rust-lang.github.io/rust-clippy/master/index.html#/wildcard_enum_match_arm) for that. The issue is that if you add a new enum variant, you should be explicit that you know about that new value and handled it, instead of silently ignored until it crashes in the runtime.

Reply:
Good point! I only did that when inside of objects where the only events are `ObjectKey` and `EndObject`. If not, there is a big bug I would love to get a bug report about it. I also fuzz this parser and the JSON parser to ensure there is no crash.

Reply:
"if we rely on fuzzing to catch crashes in Rust, we are doing something wrong"  :rofl:  (not really, but if something can be done by the compiler, might as well? well... usually :)

Reply:
> not really, but if something can be done by the compiler, might as well? well... usually

Yes, I should have found a better way to type the events generated by the JSON parser. This is a workaround of bad typing like often.

Reply:
can't blame you, me same way :)

Reply:
Oops, i meant this about the error enum variants